### PR TITLE
Set current tenant

### PR DESCRIPTION
**Pull request by mtahir-bit**

### CODE REVIEW

File: app/controllers/expenses_controller.rb [hunk 1]
```diff
@@ -64,21 +64,21 @@ def destroy
...
       end
       format.json { head :no_content }
     end
   end
 
   private
 
-  def set_current_company
-    @company = Company.find(params[:company_id])
+  def set_company
```
Comment (talhajunaid63) [inline app/controllers/expenses_controller.rb:75]: Let's move this method inside ApplicationController.

File: app/controllers/job_titles_controller.rb [hunk 2]
```diff
@@ -2,12 +2,12 @@
 
 class JobTitlesController < ApplicationController
   before_action :authenticate_user!
-  before_action :set_current_company
+  before_action :set_company
```
Comment (talhajunaid63) [inline app/controllers/job_titles_controller.rb:5]: I think set_company can also be moved inside application_controller

File: app/models/company_user.rb [hunk 3]
```diff
@@ -41,7 +43,7 @@ class CompanyUser < ApplicationRecord
   validates :salary, numericality: { greater_than: 0 }
   validates :user, uniqueness: { scope: :company_id, message: 'User already exist' }
 
-  after_commit :update_probation_end_date
+  after_create :update_probation_end_date
```
Comment (talhajunaid63) [inline app/models/company_user.rb:46]: we need to check if probation_changed? 

```ruby
if probations true then probation_ends_at = joining_date + 3.months else nil
```

File: app/models/device.rb [hunk 4]
```diff
@@ -24,6 +24,8 @@
 #  fk_rails_...  (company_id => companies.id)
 #
 class Device < ApplicationRecord
+  acts_as_tenant :company
+
   belongs_to :company
```
Comment (talhajunaid63) [inline app/models/device.rb:29]: `belongs_to :company` I think this is redundant if we are adding `acts_as_tenant`

File: app/views/clients/index.html.erb [hunk 5]
```diff
@@ -11,14 +11,16 @@
...
           </li>
-          <li>
-            <div class="flex items-center">
-              <svg class="w-6 h-6 text-gray-400" fill="currentColor" viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg">
-                <path fill-rule="evenodd" d="M7.293 14.707a1 1 0 010-1.414L10.586 10 7.293 6.707a1 1 0 011.414-1.414l4 4a1 1 0 010 1.414l-4 4a1 1 0 01-1.414 0z" clip-rule="evenodd"></path>
-              </svg>
-              <a href=<%= company_path(@company) %>  class="ml-1 text-gray-700 hover:text-primary-600 md:ml-2 dark:text-gray-300 dark:hover:text-white"><%= @company.name %></a>
-            </div>
-          </li>
+          <% if @company.present? %>
```
Comment (talhajunaid63) [inline app/views/clients/index.html.erb:14]: why we need this check `@company` should always be present regardless of the user type visiting the clients path.

File: app/views/clients/index.html.erb [hunk 6]
```diff
@@ -40,7 +42,7 @@
           </div>
         </form>
       </div>
-      <%= link_to "Add new client", new_company_client_path, data: { "turbo-frame": "remote_modal" }, class: "text-white bg-primary-700 hover:bg-primary-800 focus:ring-4 focus:ring-primary-300 font-medium rounded-lg text-sm px-5 py-2.5 dark:bg-primary-600 dark:hover:bg-primary-700 focus:outline-none dark:focus:ring-primary-800" %>
+      <%= link_to "Add new client", new_client_path, data: { "turbo-frame": "remote_modal" }, class: "text-white bg-primary-700 hover:bg-primary-800 focus:ring-4 focus:ring-primary-300 font-medium rounded-lg text-sm px-5 py-2.5 dark:bg-primary-600 dark:hover:bg-primary-700 focus:outline-none dark:focus:ring-primary-800" if @company.present? %>
```
Comment (talhajunaid63) [inline app/views/clients/index.html.erb:45]: why `if @company.present?` this check?

File: app/controllers/application_controller.rb [hunk 7]
```diff
@@ -4,13 +4,42 @@ class ApplicationController < ActionController::Base
...
-  # before_action :set_current_companyt
+  before_action :clear_company_id_from_session
+  set_current_tenant_through_filter
+  before_action :set_company_tenant
 
-  # def set_current_tenant
-  #   current_company = Company.find_by(id: params[:id])
-  #   set_current_tenant(current_company)
-  # end
+  def set_company_tenant
```
Comment (talhajunaid63) [inline app/controllers/application_controller.rb:11]: Lets move this inside a concern `CurrentRequestDetail` or something. Also, we need to add another attribute at Current class `company_user`.

```ruby
Current.company_user = CompanyUser.find_by(user: current_user, company: Current.company)
```

File: app/views/clients/index.html.erb (hunk 5 above)
Comment (mtahir-bit) [reply app/views/clients/index.html.erb:14]: for admins. admin can visit all the index pages without company present and see the data of all companies

File: app/views/clients/index.html.erb (hunk 6 above)
Comment (mtahir-bit) [reply app/views/clients/index.html.erb:45]: for admins as they can visit index pages with/without company present

File: app/controllers/home_controller.rb [hunk 8]
```diff
@@ -14,7 +14,7 @@ def index
   private
 
   def company
-    @company ||= Company.first
+    @company ||= Current.company
```
[the remark this answers is not in the document]
Comment (mtahir-bit) [reply app/controllers/home_controller.rb:17]: It shows empty plots. In the future we can update it to show combined data of all companies for admins.

**Pull request merged**